[WellSQL Migration] Migrate `WCRevenueStatsModel` to Room

ParaskP7 opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->

Closes: [AINFRA-128](https://linear.app/a8c/issue/AINFRA-128)

---

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR migrates `WCRevenueStatsModel` from [WellSQL](https://github.com/wordpress-mobile/wellsql) to [Room](https://developer.android.com/training/data-storage/room) database.

---

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

#### `My store` + `Performance`

1. Navigate on the `My store` tab and verify that the `Performance` section is being correctly populated with the expected `Revenue` number.
2. On that `Performance` section, click on the `Calendar` icon and select `Today`. Verify that the section's title changes to something like `Today Oct 2, 2025` and that the `Revenue` number is updated accordingly.
3. Similarly, select `This Week`, `This Month` and `This Year` and verify everything is working as expected.
4. Open `App Inspection`, check the `RevenueStatsEntity` table, and notice 4 rows being populated, each for every non-custom selection of yours (so far):
    - Today: `Today2025-10-022025-10-02` (rangeId)
    - This Week: `Week to Date2025-09-282025-10-04` (rangeId)
    - This Month: `Month to Date2025-10-012025-10-31` (rangeId)
    - This Year: `Year to Date2025-01-012025-12-31` (rangeId)
6. Now, select `Custom` and edit the section's title picking a specific "single" day like `Oct 2 - Oct 2`, click `SAVE`. Verify that the section's title changes to something like `Custom Oct 2, 2025 - Oct 2, 2025` and that the `Revenue` number is updated accordingly.
7. Then, edit the section's title again, picking any other range, except for a "single" day, like `Oct 1 - Oct 2`, click `SAVE`. 
Verify that the section's title changes to something like `Custom Oct 1, 2025 - Oct 2, 2025` and verify the `Revenue` number is updated accordingly.
8. Try the below custom dates:
    - Oct 2 - Oct 2 (Full Day): `Custom2025-10-022025-10-02` (rangeId)
    - Oct 1 - Oct 2 (Partial Week): `Custom2025-10-012025-10-02` (rangeId)
    - Sep 21 - Sep 27 (Full Week): `Custom2025-09-212025-09-27` (rangeId)
    - Sep 1 - Sep 30 (Full Month): `Custom2025-09-012025-09-30` (rangeId)
    - Jan 1 - Oct 2 (Partial Year): `Custom2025-01-012025-10-02` (rangeId)

FYI: I've also compared the before (`WCRevenueStatsModel`) and after (`RevenueStatsEntity`) migration data on those tables after making the exact same date selections and everything LGTM, see `Images/gif` below:

---

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

<img width="1915" height="613" alt="image" src="https://github.com/user-attachments/assets/148785c3-9f97-4e43-935d-0a6cc140832d" />

#### BEFORE `WCRevenueStatsModel`

<img width="1601" height="340" alt="BEFORE(WCRevenueStatsModel)" src="https://github.com/user-attachments/assets/7b9556a3-fa37-4ee7-9522-de97c86c6f5b" />

#### AFTER `RevenueStatsEntity`

<img width="1601" height="340" alt="AFTER(RevenueStatsEntity)" src="https://github.com/user-attachments/assets/a28beac3-cd6a-4e18-8eb9-4cef1908470d" />

---

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

ParaskP7 committed:
Cleanup: Remove unused available flag from revenue response payload

ParaskP7 committed:
Cleanup: Remove unnecessary rows affected from on revenue stats changed

ParaskP7 committed:
Refactor: Make getXyzRevenueStats suspend functions

FYI: This is done in preparation to the subsequent WellSQL to Room
table migration.

ParaskP7 committed:
Migrate: 'WCRevenueStatsModel' from wellsql to room

FYI: This change includes but is not limited to the:
- Entity Introduction
- DAO Introduction (With Test)
- WCAndroidDatabase Migration
- Code Adjustments (Store)
- Test Adjustments & Fixes (Store)

ParaskP7 committed:
Db: Drop 'WCRevenueStatsModel' table from 'wp-fluxc'

ParaskP7 committed:
Analysis: Suppress unit test naming rule closer to revenue stats test

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/model/WCRevenueStatsModel.kt:
```diff
@@ -1,37 +1,31 @@
-    @Column var total = "" // JSON - A map of total stats for a given time period
-    @Column var rangeId = "" // A ID for easily fetch the Revenue Stats for a given start and end date
-
+@Entity(
+    tableName = "RevenueStatsEntity",
+    primaryKeys = ["localSiteId", "rangeId"],
+)
+data class WCRevenueStatsModel(
```
Hmmm, you are right but:

1. During my testing I couldn't find a way to get `rangeId` come as empty string from the REST API.
2. Initially, I indeed tried the `localSiteId` + `interval` + `startDate` + `endDate` combination as primary key. But, then decided in favor of using the `localSiteId` + `rangeId` instead. This is because when testing with the former 4 keyed primary key I was getting 8/9 rows and thus couldn't replicate what I was seeing with the previous WellSQL implementation, see below:

#### BEFORE `WCRevenueStatsModel`

<img width="1601" height="340" alt="BEFORE(WCRevenueStatsModel)" src="https://github.com/user-attachments/assets/7b9556a3-fa37-4ee7-9522-de97c86c6f5b" />

#### AFTER `RevenueStatsEntity`

<img width="1601" height="340" alt="AFTER(RevenueStatsEntity)" src="https://github.com/user-attachments/assets/a28beac3-cd6a-4e18-8eb9-4cef1908470d" />

Now notice that when using the 4 keyed primary key, the 1st and 5th (`HOURS`) rows completely match, no matter if that's for a non-custom or custom date and thus data is lost, depending which you selected last. Which, to be honest is not a big deal, but nevertheless I wanted to preserve as much parity between the 2 solutions. 💭

wzieba replied:
The other option would be to stick to artificial primary key (as in `WCRevenueStatsModel`), but I think `rangeId` should be... `id` - a unique value (in context of a site).

I think your choice is correct. But you might want also check this PR (if you haven't yet) that introduces the `rangeId` and a check if it's empty or not: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2785/files#diff-3142dcb57a32d3e922e5e96b706565af858c76f424d01ecc16a41af257c37649R135-R153


> During my testing I couldn't find a way to get `rangeId` come as empty string from the REST API.

I guess that the only use case for this is when a 3rd party plugin modifies this contract. It's though to assess the risk, though.

ParaskP7 replied:
Thanks for taking a look at that @wzieba ! 🙇 

> The other option would be to stick to artificial primary key (as in WCRevenueStatsModel), but I think rangeId should be... id - a unique value (in context of a site).

I too think `rangeId` should be a unique id value. 🤷 

> I think your choice is correct. But you might want also check this PR (if you haven't yet) that introduces the rangeId and a check if it's empty or not: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2785/files#diff-3142dcb57a32d3e922e5e96b706565af858c76f424d01ecc16a41af257c37649R135-R153

I think I did, but I don't remember anymore... 🫣 Will check again and come back to you on that! 💯 

> I guess that the only use case for this is when a 3rd party plugin modifies this contract. It's though to assess the risk, though.

True, good thinking, this is not the first time we see that! 🤔

ParaskP7 replied:
👋 @wzieba !

> I think I did, but I don't remember anymore... 🫣 Will check again and come back to you on that! 💯

So, I did investigate that today and came to the conclusion that:

1. This [FluxC#2785](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2785) PR was the one introducing this `rangeId`. But, I don't see any evidence as to why it could be empty. 🤷 
2. And specific FluxC [commit](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2785/commits/6a1fccffb255bb512496773c6e883563f182b5ba) introduces this `rangeId != 0` check. Again, I don't see any evidence as to why this `rangeId` can be `0` or `empty`. 🤷 
3. I then went and check all the places this gets used and figured out a way to test it. I found that indeed this gets to be empty when it is used in a widget. ✅ (or wearable, not tested)
4. However, instead of reverting to using an artificial primary key, and make this all that bit more complicated, unnecessarily if you ask me, I chose to keep the migration as is and introduce a bit of client functionality to populate this `rangeId` with a non-empty value, following the same `Custom` pattern, using either the `Widget` or `Wear` prefix, followed by a `Start` + `End` date, the result: 5c457baec69732cb36f2572931a917aadd148c7a

<img width="1600" height="250" alt="image" src="https://github.com/user-attachments/assets/b601d068-6ef6-4888-93b3-8732b5c50910" />

Let me know what you think. 🙏 

FYI: Let me also ping a product engineer here as an extra pair of eyes on this specific change. Cc @malinajirka and thanks for helping us out here. 🙇

PS: Btw, this `rangeId` solution, even without this change here was working as expected without any crashes, because this revenue widget is quite simple, only targeting today (`HOURS`, without a means to change that (or I don't know how). 🤔

<img width="318" height="117" alt="image" src="https://github.com/user-attachments/assets/e7c33c7e-2a57-4bea-bd5b-621b87bc9573" />

wzieba replied:
Thank you @ParaskP7 ! Great research, I haven't thought about widget or wear cases. If this property is controlled by the app, and we asserted by code design that it can't be null, then it's great 👍 . I left minor comment about readability/maintainability of this: https://github.com/woocommerce/woocommerce-android/commit/5c457baec69732cb36f2572931a917aadd148c7a#r167567843

ParaskP7 replied:
Awesome, thanks for taking another look at this @wzieba , much appreciated! 🙇 ❤️ 

> If this property is controlled by the app, and we asserted by code design that it can't be null, then it's great 👍 .

👍 

> I left minor comment about readability/maintainability of this: https://github.com/woocommerce/woocommerce-android/commit/5c457baec69732cb36f2572931a917aadd148c7a#r167567843

Will do! 👍

ParaskP7 replied:
FYI: This is now done: d54dc9e667a7f7d7944dfba113fdeb89e9cbb1c4 + 0cb4a448f3cb3a3ab8bf8987521d5daac533491c + e61a21b80457a964233fca41659e0f97be579b0b Cc @wzieba 

Let me know how you like this 3 x refactors I did and I'll proceed with merging this afterward! 🙏

wzieba replied:
look cool, thank you @ParaskP7 🙌

ParaskP7 replied:
Awesome, thanks for checking on it @wzieba , will now update the branch and merge this as soon as CI goes 🟢 ! 🙇 ❤️ 🚀

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/persistence/dao/RevenueStatsDao.kt:
```diff
@@ -0,0 +1,43 @@
+
+@Dao
+internal abstract class RevenueStatsDao {
+    @Query(
+        """
+        SELECT * FROM RevenueStatsEntity
+        WHERE localSiteId = :siteId
+        AND interval = :granularity
+        AND startDate = :startDate
+        AND endDate = :endDate
+        """
+    )
```
Yea, I was thinking of adding a `@TypeConverter` for this but wasn't too sure if that's too necessary in this case, @wzieba wdty? 🤔

wzieba replied:
I think a small `@TypeConvert` could be a good decision to not leave this conversion vulnerable to changes of Kotlin or Room (e.g. upper/lower case etc.).

ParaskP7 replied:
Thanks for the reply @wzieba , convinced, will go ahead and add that small  `@TypeConvert`. 👍

ParaskP7 replied:
FYI: This is now done: 1fe478809871113362dd10b05b0d3c7403d89ee8 Cc @wzieba

ParaskP7 commented on libs/fluxc-plugin/src/main/kotlin/org/wordpress/android/fluxc/store/WCStatsStore.kt:
```diff
@@ -106,10 +107,9 @@ class WCStatsStore @Inject internal constructor(
-        val site: SiteModel,
-        val available: Boolean = false
+        val site: SiteModel
     ) : Payload<OrderStatsError>() {
-        constructor(error: OrderStatsError, site: SiteModel, available: Boolean) : this(site, available) {
+        constructor(error: OrderStatsError, site: SiteModel) : this(site) {
```
Don't worry about that, this was completed unused and thus got cleaned-up. 🧹

wzieba commented on libs/fluxc-plugin/src/test/java/org/wordpress/android/fluxc/store/WCStatsStoreTest.kt:
```diff
@@ -385,7 +376,13 @@ class WCStatsStoreTest {
+                FetchRevenueStatsResponsePayload(
+                    it,
+                    StatsGranularity.DAYS,
+                    WCRevenueStatsModel(LocalId(1), "", "", "", "", "", "")
```
nit: I've noticed `WCRevenueStatsModel(LocalId(1), "", "", "", "", "", "")` is repeated in this PR. WDYT about creating a factory, **test-only** extension method? I'm thinking about

```kotlin
fun WCRevenueStatsModel.Companion.fake() = WCRevenueStatsModel(LocalId(1), "", "", "", "", "", "")
```

and then 

```suggestion
                    WCRevenueStatsModel.fake()
```

Or you can even make it more customisable like `revenue` method from `RevenueStatsDaoTest`

ParaskP7 replied:
Thanks for the suggestion @wzieba , will do, I also like how you've named it `fake()`! 💯

ParaskP7 replied:
FYI: This is now done: 92d5229bf68060306e9fd5b6da6d0add2ea5123a Cc @wzieba

wzieba reviewed (approved):
Looks good, nice job! 👏 

<img width="874" height="229" alt="Image" src="https://github.com/user-attachments/assets/698927fd-867a-4e2c-8339-59028f855e51" />

I also couldn't make `rangeId` empty or null. Leaving some comments but nothing blocking.

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-revenue-stats-model-to-room

ParaskP7 committed:
Refactor: Make revenue range id not empty by default

It seems that 'revenueRangeId' it totally app driven, and not REST API
driven, and as such, can be completely manipulated, effectively becoming
a non-empty value at all times.

With this change, we are making sure that when 'revenueRangeId' comes
from places like 'Widget' or 'Wear', we are adding the necessary missing
information to uniquely identify this revenue range id. We are using
this '<Medium><StartDate><EndDate>' pattern.

FYI: This extra 'Medium' data was added to easily identify the medium.
With or without 'Medium' this would have worked fine anyway.

PS: Where needed, 'Medium' is either 'Widget' or 'Wear'.

------------------------------------------------------------------------

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14680#discussion_r2398949681

ParaskP7 committed:
Migration: Introduce type converter for stats granularity

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14680#discussion_r2398949700

ParaskP7 committed:
Test: Reduce duplicate fakes within wc stats store test

PR Comment: https://github.com/woocommerce/woocommerce-android/pull/
14680#discussion_r2407770941

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-revenue-stats-model-to-room

ParaskP7 committed:
Analysis: Fix detect issues on type converter tests

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-revenue-stats-model-to-room

malinajirka reviewed (approved):
Thanks @ParaskP7 ! It all looks good to me. I've reviewed the commits as well as tested the app and I haven't found any issues. I've compared the before and after in a split view on UI as well as with App inspector and the data in DB look as expected.


> FYI: We are on a meetup next week,

Looking forward to your visit!!!! 

>  then sabbatical

🥳 🍾 🎆 🥳 🍾 🎆 🥳 🍾 🎆 🥳 🍾 🎆 🥳 🍾 🎆 🥳 🍾 🎆

ParaskP7 committed:
Refactor: Move stats time range data class to libs commons

FYI: This removes 'StatsTimeRange' from both, the 'WooCommerce' and
'WooCommerce-Wear' modules.

ParaskP7 committed:
Refactor: Move to revenue range id extension function to libs commons

FYI: This removes 'toRevenueRangeId()' from both, the 'WooCommerce' and
'WooCommerce-Wear' modules.

ParaskP7 committed:
Refactor: Move as revenue range id extension function to libs commons

FYI: This removes 'toRevenueRangeId()' from the 'WooCommerce' modules.

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-revenue-stats-model-to-room

ParaskP7 committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce-android into migrate/wc-revenue-stats-model-to-room
